drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2351 @@
+	int ret;
+
+	if (sys_read32(config->regs + HW_CAPABILITY) & HW_CAPABILITY_SLV_HJ_CAP) {
+		LOG_ERR("%s: HJ not supported", dev->name);
+		return -ENOTSUP;
+	}
+	if (sys_read32(config->regs + DEVICE_ADDR) & DEVICE_ADDR_DYNAMIC_ADDR_VALID) {
+		LOG_ERR("%s: HJ not available, DA already assigned", dev->name);
+		return -EACCES;
+	}
+	/* if this is set, then it is assumed it is already trying */
+	if ((sys_read32(config->regs + SLV_EVENT_STATUS) & SLV_EVENT_STATUS_HJ_EN)) {
+		LOG_ERR("%s: HJ requests are currently disabled by DISEC", dev->name);
+		return -EAGAIN;
+	}
+
+	/*
+	 * This is issued auto-magically by the IP when certain conditions are meet.
+	 * These include:
+	 * 1. SLV_EVENT_STATUS[HJ_EN] = 1 (or a controller issues Enables HJ events with
+	 * the CCC ENEC, This can be set to 0 with CCC DISEC from a controller)
+	 * 2. The Dynamic address is invalid. (not assigned yet)
+	 * 3. Bus Idle condition is met (1ms) as programmed in the Bus Timing Register
+	 */
+
+	/* enable HJ */
+	sys_write32(sys_read32(config->regs + SLV_EVENT_STATUS) | SLV_EVENT_STATUS_HJ_EN,
+		    config->regs + SLV_EVENT_STATUS);
+
+	ret = k_sem_take(&data->sem_hj, K_MSEC(1000));
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int dw_i3c_target_ibi_raise_tir(const struct device *dev, struct i3c_ibi *request)
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	int status;
+	uint32_t slv_intr_req, slv_ibi_resp;
+
+	if (sys_read32(config->regs + HW_CAPABILITY) & HW_CAPABILITY_SLV_IBI_CAP) {
+		LOG_ERR("%s: IBI TIR not supported", dev->name);
+		return -ENOTSUP;
+	}
+
+	if (!(sys_read32(config->regs + DEVICE_ADDR) & DEVICE_ADDR_DYNAMIC_ADDR_VALID)) {
+		LOG_ERR("%s: IBI TIR not available, DA not assigned", dev->name);
+		return -EACCES;
+	}
+
+	if (!(sys_read32(config->regs + SLV_EVENT_STATUS) & SLV_EVENT_STATUS_SIR_EN)) {
+		LOG_ERR("%s: IBI TIR requests are currently disabled by DISEC", dev->name);
+		return -EAGAIN;
+	}
+
+	slv_intr_req = sys_read32(config->regs + SLV_INTR_REQ);
+	if (sys_read32(config->regs + SLV_CHAR_CTRL) & SLV_CHAR_CTRL_IBI_PAYLOAD) {
+		uint32_t tir_data = 0;
+
+		/* max support length is DA + MDB (1 byte) + 4 data bytes, MDB must be at least
+		 * included
+		 */
+		if ((request->payload_len > 5) || (request->payload_len == 0)) {
+			return -EINVAL;
+		}
+
+		/* MDB should be the first byte of the payload */
+		slv_intr_req |= SLV_INTR_REQ_MDB(request->payload[0]) |
+				SLV_INTR_REQ_SIR_DATA_LENGTH(request->payload_len - 1);
+
+		/* program the tir data packet */
+		tir_data |=
+			SLV_SIR_DATA_BYTE0((request->payload_len > 1) ? request->payload[1] : 0);
+		tir_data |=
+			SLV_SIR_DATA_BYTE1((request->payload_len > 2) ? request->payload[2] : 0);
+		tir_data |=
+			SLV_SIR_DATA_BYTE2((request->payload_len > 3) ? request->payload[3] : 0);
+		tir_data |=
+			SLV_SIR_DATA_BYTE3((request->payload_len > 4) ? request->payload[4] : 0);
+		sys_write32(config->regs + SLV_SIR_DATA, tir_data);
```

Comment:
Must be sys_write32(tir_data, config->regs + SLV_SIR_DATA);

Reply:
done